src/snowflake/cli/_plugins/nativeapp/sf_sql_facade.py

```diff
@@ -0,0 +1,223 @@
+
+
+class UnknownSQLError(Exception):
+    """Exception raised when the root of the SQL error is unidentified by us."""
+
+    # PJ-question how do we ensure exit codes remain unique
+    exit_code = 3
+
+    def __init__(self, message):
+        msg = f"Unknown SQL error occurred. {message}"
+        super().__init__(msg)
+        self.message = msg
+
+    def __str__(self):
+        return self.message
+
+
+class UserScriptError(ClickException):
+    """Exception raised when user-provided scripts fail."""
+
+    def __init__(self, script_name, msg):
+        super().__init__(f"Failed to run script {script_name}. {msg}")
+
+
+class SnowflakeSQLFacade:
+    def __init__(self, sql_executor: SqlExecutor | None):
+        self._sql_executor = (
+            sql_executor if sql_executor is not None else get_sql_executor()
+        )
+
+    @cached_property
+    def _log(self):
+        return logging.getLogger(__name__)
+
+    @contextmanager
+    def _use_warehouse_optional(self, new_wh: str | None):
+        """
+        Switches to a different warehouse for a while, then switches back.
+        This is a no-op if the requested warehouse is already active or if no warehouse is passed in.
+        @param new_wh: Name of the warehouse to use. If not a valid Snowflake identifier, will be converted before use.
+        """
+        if new_wh is None:
+            yield
+            return
+
+        valid_wh_name = to_identifier(new_wh)
+
+        wh_result = self._sql_executor.execute_query(
+            f"select current_warehouse()"
+        ).fetchone()
+
+        # If user has an assigned default warehouse, prev_wh will contain a value even if the warehouse is suspended.
+        try:
+            prev_wh = wh_result[0]
+        except:
+            prev_wh = None
+        # new_wh is not None, and should already be a valid identifier, no additional check is performed here.
+        is_different_wh = valid_wh_name != prev_wh
+        if is_different_wh:
+            self._log.debug(f"Using warehouse: {valid_wh_name}")
+            try:
+                self._sql_executor.execute_query(f"use warehouse {valid_wh_name}")
+            except ProgrammingError as err:
+                if err.errno == DOES_NOT_EXIST_OR_CANNOT_BE_PERFORMED:
+                    raise CouldNotUseObjectError(
+                        ObjectType.WAREHOUSE, valid_wh_name
+                    ) from err
+
+                raise ProgrammingError(
+                    f"Failed to use warehouse {valid_wh_name}"
+                ) from err
+            except Exception as err:
+                raise UnknownSQLError(
+                    f"Failed to use warehouse {valid_wh_name}"
+                ) from err
+        try:
+            yield
+        finally:
+            if is_different_wh and prev_wh is not None:
+                self._log.debug(f"Switching back to warehouse:{prev_wh}")
+                self._sql_executor.execute_query(f"use warehouse {prev_wh}")
```

Comment:
ditto, what happens if this fails? Technically the user script could drop the warehouse or otherwise make it inaccessible.

Reply:
good point. and the same can be said about the `use database` case. See update. made a `_use_object` method that has the error handling it and using it for all `use object_type object_name` calls.